### PR TITLE
specification: CoVE-IO theory of operations

### PR DESCRIPTION
**Pull request by sameo**
Describe CoVE-IO ABI operations, including platform setup, device initialization and interface assignment.

### CODE REVIEW

**Comment by sameo** [comment]
cc @rsahita

File: specification/theory_operations.adoc [hunk 1]
```diff
@@ -1,2 +1,282 @@
 [[theory_operations]]
+
 == Theory of Operations
+
+The CoVE-IO specification extends the CoVE host and guest SBI extensions to
+allow TVMs to establish trust with TEE-IO devices, and then use and interact
+with those devices. Untrusted domain components are responsible for assigning
```
Comment (rsahita) [inline specification/theory_operations.adoc:7]:
```suggestion
with those devices. Untrusted supervisor domain components are responsible for assigning
```

File: specification/theory_operations.adoc [hunk 2]
```diff
@@ -1,2 +1,282 @@
...
+TEE-IO devices to TVMs, and also for supporting the TVM acceptance or rejection
+of the assigned devices. The TSM, on the other hand, establishes and maintains
+the secure physical and logical links between TVMs and their assigned devices.
+
+The following sections describe the functionality of the TSM provided CoVE-IO
```
Comment (rsahita) [inline specification/theory_operations.adoc:12]:
```suggestion
The following sections describe the functionality of the TSM-provided CoVE-IO
```

File: specification/theory_operations.adoc [hunk 3]
```diff
@@ -1,2 +1,277 @@
...
+   notifies the TSM about a pending trusted IOMMU MSI by calling the
+   `sbi_covh_notify_iommu_msi()` COVH function.
+7. The TSM verifies that there is a pending MSI by reading the IOMMU `s_ipsr`
+   register, and handles the interrupts as needed.
+
+The untrusted driver can trick the TSM by either invoking the
+`sbi_covh_notify_iommu_msi()` COVH function while there are no pending MSI from
+the trusted IOMMU or not invoking it when there actually is one such pending
+interrupt. In the former case, the TSM can verify that there really is a pending
+MSI by checking the trusted IOMMU status registers. The latter case could cause
```
Comment (rsahita) [inline specification/theory_operations.adoc:55]: have we introduced what we mean by "trusted IOMMU status registers"?

File: specification/theory_operations.adoc [hunk 4]
```diff
@@ -1,2 +1,277 @@
...
+a denial of service, which is not covered by the confidential computing threat
```
Comment (rsahita) [inline specification/theory_operations.adoc:56]:
```suggestion
a denial of service, which is not in scope of the confidential computing threat
```

File: specification/theory_operations.adoc [hunk 5]
```diff
@@ -1,2 +1,277 @@
...
+%%{init: {'theme': 'neutral', 'themeVariables': {'darkMode': true}, "flowchart" : { "curve" : "basis" } } }%%
+sequenceDiagram
+
+autonumber
+
+participant VMM as Untrusted Domain Manager (VMM)
+participant TSM
+participant DSM as Device DSM or RoT
+
+VMM ->> TSM: [COVH] - sbi_covh_create_spdm_session()
```
Comment (jyao1) [inline specification/theory_operations.adoc:226]: It is a good start.

We may consider how to enable IDE e.g. IDE_KM protocol.

File: specification/theory_operations.adoc [hunk 6]
```diff
@@ -1,2 +1,277 @@
...
+
+For each TDI that gets attached to a TVM, its corresponding PCIe root port must
+be configured with per-device IDE keys. This configuration is done through the
+PCIe IDE key management protocol (IDE_KM) that runs on top of an SPDM session.
+It is thus necessary to establish an SPDM session with all trusted I/O
+registered root ports. This is akin to considering the RP as just another
+trusted device for which the DSM is the platform RoT. The SPDM establishment
+thus happens between the TSM and the platform RoT. By having the RoT playing the
+RP DSM role, the RP vendor-specific IDE key management interface is abstracted
+by the RoT.
```
Comment (rsahita) [inline specification/theory_operations.adoc:104]: do we have a figure for this?

File: specification/theory_operations.adoc [hunk 7]
```diff
@@ -1,2 +1,278 @@
...
+VMM ->> DSM: [DOE] - SPDM_FINISH
+DSM ->> VMM: [DOE] - FINISH_RESP
+VMM ->> TSM: [COVH] - spdm_resp(FINISH_RESP)
+
+Note over VMM,DSM: TSM: SPDM Session Created
+
+TSM ->> VMM: [COVH] - spdm_session_created()
+....
+
+=== Device Initialization
```
Comment (jyao1) [inline specification/theory_operations.adoc:276]: I feel SPDM session should be part of "Device Initialization".

"Platform Initialization" should handle host side only (such as IOMMU), but not device side.

**Review by rsahita — CHANGES_REQUESTED** [review]
some comments and change suggestions

File: specification/theory_operations.adoc [hunk 8]
```diff
@@ -1,2 +1,278 @@
...
+
+Note over VMM,DSM: TSM: SPDM Connection Setup
+
+TSM ->> TSM: Generate SPDM request: GET_VERSION
+TSM ->> VMM: [COVH] - spdm_req(GET_VERSION)
+VMM ->> DSM: [DOE] - SPDM_GET_VERSION
+DSM ->> VMM: [DOE] - SPDM_VERSION
+VMM ->> TSM: [COVH] - spdm_resp(VERSION)
+
+TSM ->> TSM: Generate SPDM request: GET_CAPS
```
Comment (jyao1) [inline specification/theory_operations.adoc:237]: I recommend we use the name aligning with SPDM specification. E.g.
GET_CAPS => GET_CAPABILITIES
CAPS => CAPABILITIES
NEGOTIATE_ALGS => NEGOTIATE_ALGORITHMS
ALGS => ALGORITHMS
KEY_RESP => KEY_EXCHANGE_RSP
FINISH_RESP => FINISH_RSP

File: specification/theory_operations.adoc [hunk 9]
```diff
@@ -1,2 +1,278 @@
...
+DSM ->> VMM: [DOE] - KEY_RESP
+VMM ->> TSM: [COVH] - spdm_resp(KEY_RESP)
+TSM ->> TSM: Derive DHE secret
+TSM ->> TSM: Generate SPDM request: FINISH
+TSM ->> VMM: [COVH] - spdm_req(FINISH)
+VMM ->> DSM: [DOE] - SPDM_FINISH
+DSM ->> VMM: [DOE] - FINISH_RESP
+VMM ->> TSM: [COVH] - spdm_resp(FINISH_RESP)
+
+Note over VMM,DSM: TSM: SPDM Session Created
```
Comment (jyao1) [inline specification/theory_operations.adoc:271]: When do you think we need to call SPDM GET_MEASUREMENTS?

In this API, or separate API?

File: specification/theory_operations.adoc [hunk 10]
```diff
@@ -1,2 +1,282 @@
...
+
+The CoVE-IO specification extends the CoVE host and guest SBI extensions to
+allow TVMs to establish trust with TEE-IO devices, and then use and interact
+with those devices. Untrusted domain components are responsible for assigning
+TEE-IO devices to TVMs, and also for supporting the TVM acceptance or rejection
+of the assigned devices. The TSM, on the other hand, establishes and maintains
+the secure physical and logical links between TVMs and their assigned devices.
+
+The following sections describe the functionality of the TSM provided CoVE-IO
+extensions to support trusted I/O on CoVE enabled platforms.
```
[the remark this answers is not in the document]
Comment (sameo) [reply specification/theory_operations.adoc]: Done, thanks.

File: specification/theory_operations.adoc [hunk 11]
```diff
@@ -1,2 +1,277 @@
...
+
+==== IOMMU Registration and Setup
+
+The TSM relies on the availability of at least one IOMMU instance exclusvely
+associated with the TSM trusted domain. Those IOMMUs allow the TSM to perform
+address translations and protection from DMA and interrupts originating from
+assigned TDIs. Only the TSM can access and program he trusted domain IOMMU
+instances.
+
+Trusted domain assigned IOMMUs can generate MSIs in order to signal the TSM
```
[the remark this answers is not in the document]
Comment (sameo) [reply specification/theory_operations.adoc]: Done, thanks.

File: specification/theory_operations.adoc [hunk 12]
```diff
@@ -1,2 +1,277 @@
...
+about command completions, transaction faults or device page requests. Those
+MSIs target system physical memory, which is owned by the untrusted domain
+manager, e.g. the host VMM. As a consequence, it is the untrusted domain
+responsibility to reserve the MSI addresses and then rely on the TSM to program
+the IOMMUs with those reserved addresses. This IOMMU registration process is
+driven by the untrusted domain manager for all IOMMUs that participate in TEE-IO
```
[the remark this answers is not in the document]
Comment (sameo) [reply specification/theory_operations.adoc]: Done, thanks.

File: specification/theory_operations.adoc [hunk 13]
```diff
@@ -1,2 +1,277 @@
...
+   the `sbi_covh_register_iommu()` COVH function. The TSM gets the allocated MSI
+   vectors and configures the trusted IOMMU `s_msi_cfg_table` register
+   accordingly.
+6. When a trusted IOMMU sends an MSI, the untrusted IOMMU driver handles it and
+   notifies the TSM about a pending trusted IOMMU MSI by calling the
+   `sbi_covh_notify_iommu_msi()` COVH function.
+7. The TSM verifies that there is a pending MSI by reading the IOMMU `s_ipsr`
+   register, and handles the interrupts as needed.
+
+The untrusted driver can trick the TSM by either invoking the
```
[the remark this answers is not in the document]
Comment (sameo) [reply specification/theory_operations.adoc]: Done, thanks.

File: specification/theory_operations.adoc [hunk 14]
```diff
@@ -1,2 +1,277 @@
...
+manager, e.g. the host VMM. As a consequence, it is the untrusted domain
+responsibility to reserve the MSI addresses and then rely on the TSM to program
+the IOMMUs with those reserved addresses. This IOMMU registration process is
+driven by the untrusted domain manager for all IOMMUs that participate in TEE-IO
+and gies as follows:
+
+1. The TSM is loaded and provisioned with a CoVE-IO manifest.
+2. The untrusted domain manager (e.g. the host VMM) enumerates all platform
+   IOMMUs as PCIe devices.
+3. The untrusted IOMMU driver is loaded and probed.
```
[the remark this answers is not in the document]
Comment (sameo) [reply specification/theory_operations.adoc]: Agreed. Done now, thanks.

File: specification/theory_operations.adoc (hunk 3 above)
Comment (sameo) [reply specification/theory_operations.adoc]: We have not, I'll rephrase.

File: specification/theory_operations.adoc (hunk 4 above)
Comment (sameo) [reply specification/theory_operations.adoc]: Done, thanks.

File: specification/theory_operations.adoc (hunk 8 above)
Comment (sameo) [reply specification/theory_operations.adoc]: Done, thanks.

File: specification/theory_operations.adoc [hunk 15]
```diff
@@ -1,2 +1,277 @@
...
+instances.
+
+Trusted domain assigned IOMMUs can generate MSIs in order to signal the TSM
+about command completions, transaction faults or device page requests. Those
+MSIs target system physical memory, which is owned by the untrusted domain
+manager, e.g. the host VMM. As a consequence, it is the untrusted domain
+responsibility to reserve the MSI addresses and then rely on the TSM to program
+the IOMMUs with those reserved addresses. This IOMMU registration process is
+driven by the untrusted domain manager for all IOMMUs that participate in TEE-IO
+and gies as follows:
```
[the remark this answers is not in the document]
Comment (sameo) [reply specification/theory_operations.adoc]: Done, thanks.

File: specification/theory_operations.adoc [hunk 16]
```diff
@@ -1,2 +1,430 @@
...
+Those MSIs target system physical memory, which is owned by the host security
+domain manager, e.g. the host VMM. As a consequence, it is the host security
+domain's responsibility to reserve the MSI addresses and then rely on the TSM to
+program the IOMMUs with those reserved addresses. This IOMMU registration
+process is driven by the untrusted domain manager for all IOMMUs that
+participate in TEE-IO and operates as described in the following steps:
+
+1. The TSM is loaded into a supervisor domain and provisioned with a CoVE-IO
+   manifest. It is recommended that the TSM is measured by the platform
+   root-of-trust for subsequent attestation.
```
Comment (jyao1) [inline specification/theory_operations.adoc:39]: Question: Why TSM is measured by platform RoT, but not silicon RoT (CPU)?

File: specification/theory_operations.adoc [hunk 17]
```diff
@@ -1,2 +1,430 @@
...
+maintains the secure physical and logical links between TVMs and their assigned
+devices.
+
+The following sections describe the functionality of the TSM-provided CoVE-IO
+extensions to support trusted I/O on CoVE-enabled platforms.
+
+=== Platform Initialization
+
+==== IOMMU Registration and Setup
+
```
Comment (sameo) [inline specification/theory_operations.adoc:19]: Add non-normative sections related to IOMMU spec

File: specification/theory_operations.adoc [hunk 18]
```diff
@@ -1,2 +1,430 @@
...
+protection for TLPs received and transmitted between the physical device the TDI
+belongs to and its assigned PCIe root port (RP). Both endpoints (The RP and the
+physical device) must be configured with the same encryption keys through the
+IDE key management protocols and have their IDE PCIe extended capability
+configured as well.
+
+As described in the threat model and requirement sections, the following trusted
+I/O rules apply:
+
+* As PCIe switches are outside of the TVM TCB, only selective IDE streams are
```
Comment (sameo) [inline specification/theory_operations.adoc:80]: Do we need to program the switches when using selective IDE streams?

**Comment by jyao1** [comment]
General question: Do we also need consider other flow, such as interface removal? key refresh?

File: specification/theory_operations.adoc [hunk 19]
```diff
@@ -1,2 +1,430 @@
...
+  used to protect the PCIe link between the TVM and the TDI.
+* A single selective IDE stream is established between the TSM and the physical
```
[the remark this answers is not in the document]
Comment (sameo) [reply specification/theory_operations.adoc]: Fixed, thanks.

File: specification/coveio_abi.adoc [hunk 20]
```diff
@@ -1,2 +1,46 @@
...
+
+[#sbi_covh_connect_device()]
+=== Device Connection (FID TBD)
+[source, C]
+-----
+struct sbiret sbi_covh_connect_device()(unsigned long device_id,
+                                        unsigned long stream_id);
+-----
+
+Establish a secure SPDM session and configure a PCIe IDE link between the TSM
```
[the remark this answers is not in the document]
Comment (sameo) [reply specification/coveio_abi.adoc]: Correct, thanks for catching that one. Fixed now.

**Comment by sameo** [comment]
> General question: Do we also need consider other flow, such as interface removal? key refresh?

Yes, good call. I had Interface unassignment and Device disconnection in mind. Others:

- IDE key referesh
- Interface removal
- Device removal

What else?

**Comment by jyao1** [comment]
> What else?

Such SPDM key update, SPDM heartbeat.

File: specification/theory_operations.adoc (hunk 16 above)
Comment (rsahita) [reply specification/theory_operations.adoc:39]: perhaps we should say SoC RoT?

Comment (jyao1) [reply specification/theory_operations.adoc:39]: I think it depends on who is Root-of-Trust for measurement (RTM), and who loads TSM.
If TSM is loaded by host CPU, the SoC RoT makes more sense.
Unless TSM is really loaded by platform RoT, then platform RoT needs to measure it. And the platform RoT also needs to guarantee that the one CPU executing on is the same as the one loaded and measured by Platform RoT.

If we support both use case, then we can generalize it by saying: TSM is measured by Root-of-Trust for measurement (RTM). Maybe...

File: specification/theory_operations.adoc (hunk 18 above)
Comment (jyao1) [reply specification/theory_operations.adoc:80]: For example, IDE has register - "IDE Control Register (Offset 08h)"
BIT2 - **Flow-Through IDE Stream Enable** - For Switch Ports and Root Ports, Enables the Port for flow-through operation of TLPs associated with Selective IDE streams.

**Comment by jyao1** [comment]
I think it is good enough to merge now.

The rest of feedback can be tracked in other issues.

**Review by jyao1 — APPROVED** [review]

**Comment by sameo** [comment]
@rsahita I think I addressed most of your comments, so I'll go ahead and merge that PR. It's incomplete, so we'll open new issues for all the flows we want to document in this section and then we'll complete it step by step.

**Pull request merged**